### PR TITLE
LED state detection with OpenCV

### LINKED ISSUE #2: Automate LED state detection with OpenCV
## Hardware requirements
- [x] Cover resistors on CM4 with foam rubber to prevent reflection
## Software requirements
Led detection with...
- [x] uncalibrated camera, fixed position
- [x] calibrated camera, fixed position
- [x] calibrated camera with apriltag, any position

### PR DESCRIPTION
**Pull request by 4c3y**
- Created time_stamper_cv catkin package
- Implemented LED state detection with OpenCV
- Implemented visualization
- Configurable from ROS
- Unittests

### CODE REVIEW

File: time_stamper_cv/src/Node.cpp
Comment (michaelpantic) [inline time_stamper_cv/src/Node.cpp]: Please rename Node to a more verbose name

File: time_stamper_cv/src/Node.cpp
Comment (michaelpantic) [inline time_stamper_cv/src/Node.cpp]: no need for return value

File: time_stamper_cv/src/Node.cpp
Comment (michaelpantic) [inline time_stamper_cv/src/Node.cpp]: maybe rename calibration_

File: time_stamper_cv/include/Calibration.h
Comment (michaelpantic) [inline time_stamper_cv/include/Calibration.h]: do it or remove todo

File: time_stamper_cv/include/Calibration.h
Comment (michaelpantic) [inline time_stamper_cv/include/Calibration.h]: bracket initializers everywhere

File: time_stamper_cv/include/Common.h
Comment (michaelpantic) [inline time_stamper_cv/include/Common.h]: extend comment as to why we need them

File: time_stamper_cv/src/Calibration.cpp
Comment (michaelpantic) [inline time_stamper_cv/src/Calibration.cpp]: use shared_ptr for all ptrs

File: time_stamper_cv/src/Calibration.cpp
Comment (michaelpantic) [inline time_stamper_cv/src/Calibration.cpp]: increase cohesion and decrease coupling by moving shapestatus to convex_shape

File: time_stamper_cv/src/Calibration.cpp
Comment (michaelpantic) [inline time_stamper_cv/src/Calibration.cpp]: potentially wrap detector in your own class and keep keypoints and their status there

File: time_stamper_cv/src/Calibration.cpp
Comment (michaelpantic) [inline time_stamper_cv/src/Calibration.cpp]: document why we get the homography and what we are doing with it

File: time_stamper_cv/src/Calibration.cpp
Comment (michaelpantic) [inline time_stamper_cv/src/Calibration.cpp]: what is number?

Comment (michaelpantic) [reply time_stamper_cv/src/Calibration.cpp]: ah! move declaration closer to usage

Comment (michaelpantic) [reply time_stamper_cv/src/Calibration.cpp]: maybe processImage should return the number? or a result object

File: time_stamper_cv/src/Calibration.cpp
Comment (michaelpantic) [inline time_stamper_cv/src/Calibration.cpp]: add this as a custom cast converter to PointVector

File: time_stamper_cv/src/Calibration.cpp
Comment (michaelpantic) [inline time_stamper_cv/src/Calibration.cpp]: isShapeValid() != isLastShapeValid to detect change

and then construct string based on isShapeValid()

Comment (michaelpantic) [reply time_stamper_cv/src/Calibration.cpp]: (same for keypoints)

File: time_stamper_cv/src/Calibration.cpp
Comment (michaelpantic) [inline time_stamper_cv/src/Calibration.cpp]: is needed?

maybe cv::waitKey(0) or 1

File: time_stamper_cv/include/ConvexShape.h
Comment (michaelpantic) [inline time_stamper_cv/include/ConvexShape.h]: add explanation how this class works (the sorting etc)

File: time_stamper_cv/src/ConvexShape.cpp
Comment (michaelpantic) [inline time_stamper_cv/src/ConvexShape.cpp]: better: inline function not assigned with class or in some common class

File: time_stamper_cv/src/ConvexShape.cpp
Comment (michaelpantic) [inline time_stamper_cv/src/ConvexShape.cpp]: int vs double

Comment (michaelpantic) [reply time_stamper_cv/src/ConvexShape.cpp]: naming: Filter in robotics is very often used, so better something about "InRange" or "Bounded"

File: time_stamper_cv/src/ConvexShape.cpp
Comment (michaelpantic) [inline time_stamper_cv/src/ConvexShape.cpp]: use linear shift instead of multiplication

Comment (michaelpantic) [reply time_stamper_cv/src/ConvexShape.cpp]: rename to scaling
Maybe use scaling matrix?

File: time_stamper_cv/src/ConvexShape.cpp
Comment (michaelpantic) [inline time_stamper_cv/src/ConvexShape.cpp]: add comment

File: time_stamper_cv/src/ConvexShape.cpp
Comment (michaelpantic) [inline time_stamper_cv/src/ConvexShape.cpp]: add comment 
(something like "pattern not found")

File: time_stamper_cv/src/ConvexShape.cpp
Comment (michaelpantic) [inline time_stamper_cv/src/ConvexShape.cpp]: could be solved using autocorrelation

File: time_stamper_cv/src/LedParser.cpp
Comment (michaelpantic) [inline time_stamper_cv/src/LedParser.cpp]: comment on idea and kernel

File: time_stamper_cv/src/LedParser.cpp
Comment (michaelpantic) [inline time_stamper_cv/src/LedParser.cpp]: use casts

File: time_stamper_cv/src/LedParser.cpp
Comment (michaelpantic) [inline time_stamper_cv/src/LedParser.cpp]: use cv::point

File: time_stamper_cv/src/LedParser.cpp
Comment (michaelpantic) [inline time_stamper_cv/src/LedParser.cpp]: use cv::point

Comment (michaelpantic) [reply time_stamper_cv/src/LedParser.cpp]: dont forget to check against 0/0

File: time_stamper_cv/src/LedParser.cpp
Comment (michaelpantic) [inline time_stamper_cv/src/LedParser.cpp]: clean comment

File: time_stamper_cv/src/LedParser.cpp
Comment (michaelpantic) [inline time_stamper_cv/src/LedParser.cpp]: add normalization directly to kernel

File: time_stamper_cv/src/LedParser.cpp
Comment (michaelpantic) [inline time_stamper_cv/src/LedParser.cpp]: min_brightness could be a float

File: time_stamper_cv/src/Trigonometry.cpp
Comment (michaelpantic) [inline time_stamper_cv/src/Trigonometry.cpp]: try using cv::norm(a-b)

File: time_stamper_cv/src/Trigonometry.cpp
Comment (michaelpantic) [inline time_stamper_cv/src/Trigonometry.cpp]: name distance_ab

File: time_stamper_cv/src/Trigonometry.cpp
Comment (michaelpantic) [inline time_stamper_cv/src/Trigonometry.cpp]: check if a or b are 0 (avoid div/0)

File: time_stamper_cv/test/test_convex_shape.cc
Comment (michaelpantic) [inline time_stamper_cv/test/test_convex_shape.cc]: TestLowerTolerance

**Review by michaelpantic — CHANGES_REQUESTED** [review]
Looks pretty good!
- use const/ref wherever possible
- change all functionnames to be lower camel case (first letter)
- change file names to google standard
- create UML diagrams with dependencies

File: time_stamper_cv/include/common.h
Comment (michaelpantic) [inline time_stamper_cv/include/common.h]: rename to inRange or so

File: time_stamper_cv/include/configuration.h
Comment (michaelpantic) [inline time_stamper_cv/include/configuration.h]: set default value

File: time_stamper_cv/include/convex_shape.h
Comment (michaelpantic) [inline time_stamper_cv/include/convex_shape.h]: set default value

File: time_stamper_cv/include/detector.h
Comment (michaelpantic) [inline time_stamper_cv/include/detector.h]: Class names should be capitalized CamelCase

Comment (michaelpantic) [reply time_stamper_cv/include/detector.h]: (name should probably be KeyPointDetector or so)

File: time_stamper_cv/include/detector.h
Comment (michaelpantic) [inline time_stamper_cv/include/detector.h]: includes itself

File: time_stamper_cv/include/led_parser.h
Comment (michaelpantic) [inline time_stamper_cv/include/led_parser.h]: :-) Can be called homogenous 2d coordinates

File: time_stamper_cv/src/convex_shape.cpp
Comment (michaelpantic) [inline time_stamper_cv/src/convex_shape.cpp]: this hsould probably be isWithinTolerance or similar

File: time_stamper_cv/src/detector.cpp
Comment (michaelpantic) [inline time_stamper_cv/src/detector.cpp]: why not just return a string?

File: time_stamper_cv/src/led_parser.cpp
Comment (michaelpantic) [inline time_stamper_cv/src/led_parser.cpp]: hmm LedStateDetector would be a more verbose name

File: time_stamper_cv/include/common.h
Comment (michaelpantic) [reply time_stamper_cv/include/common.h]: params should all be const

File: time_stamper_cv/include/common.h
Comment (michaelpantic) [inline time_stamper_cv/include/common.h:27]: param const

File: time_stamper_cv/include/convex_shape.h
Comment (michaelpantic) [inline time_stamper_cv/include/convex_shape.h:21]: param const

File: time_stamper_cv/include/convex_shape.h
Comment (michaelpantic) [inline time_stamper_cv/include/convex_shape.h:47]: param const

File: time_stamper_cv/include/convex_shape.h
Comment (michaelpantic) [inline time_stamper_cv/include/convex_shape.h]: params const

File: time_stamper_cv/include/image_processor.h
Comment (michaelpantic) [inline time_stamper_cv/include/image_processor.h:28]: param const

File: time_stamper_cv/include/led_parser.h
Comment (michaelpantic) [inline time_stamper_cv/include/led_parser.h]: params ocnst

**Review by michaelpantic — APPROVED** [review]
Looks good to me!

See comments - please mark them as resolved (if resolved) or comment on them before hitting the merge button :-)

Also - make sure it passes the build before merging

File: time_stamper_cv/src/detector.cpp
Comment (4c3y) [reply time_stamper_cv/src/detector.cpp]: Because the std::function is intended to only get called if the keypoint state change.

File: time_stamper_cv/include/common.h
Comment (4c3y) [reply time_stamper_cv/include/common.h:27]: This can't be const because otherwise the unary operator fails to compile

File: time_stamper_cv/include/convex_shape.h
Comment (4c3y) [reply time_stamper_cv/include/convex_shape.h:47]: Added const in cpp file to avoid warnings

File: time_stamper_cv/include/convex_shape.h
Comment (4c3y) [reply time_stamper_cv/include/convex_shape.h:21]: Added const in cpp file to avoid warnings

File: time_stamper_cv/include/convex_shape.h
Comment (4c3y) [reply time_stamper_cv/include/convex_shape.h]: same here

File: time_stamper_cv/include/image_processor.h
Comment (4c3y) [reply time_stamper_cv/include/image_processor.h:28]: same here

File: time_stamper_cv/include/led_parser.h
Comment (4c3y) [reply time_stamper_cv/include/led_parser.h]: same here

File: time_stamper_cv/src/led_parser.cpp
Comment (4c3y) [reply time_stamper_cv/src/led_parser.cpp]: Renamed to LedStateParser to not confuse it with KeypointDetector

**Pull request merged**
